### PR TITLE
[SPARK-47839][SQL] Fix aggregate bug in RewriteWithExpression

### PR DESCRIPTION
**Pull request by kelvinjian-db**
### What changes were proposed in this pull request?

- Fixes a bug where `RewriteWithExpression` can rewrite an `Aggregate` into an invalid one. The fix is done by separating out the "result expressions" from the "aggregate expressions" in the `Aggregate` node, and rewriting them separately.
- Some QOL improvements around `With`:
  - Fix aliases created by `With` expression to use the `CommonExpressionId` to avoid duplicate aliases (added a conf to fall back to old behaviour, which is useful to keep the IDs consistent for golden files tests)
  - Implemented `QueryPlan.transformUpWithSubqueriesAndPruning` that the new logic depends on

### Why are the changes needed?

See [JIRA ticket](https://issues.apache.org/jira/browse/SPARK-47839) for more details on the bug that this fixes.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added new unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by kelvinjian-db** [comment]
cc @cloud-fan @jchen5

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala [hunk 1]
```diff
@@ -21,36 +21,57 @@ import scala.collection.mutable
...
+        applyInternal(p)
+    }
+  }
+
+  private def applyInternal(p: LogicalPlan): LogicalPlan = {
+    val inputPlans = p.children.toArray
+    var newPlan: LogicalPlan = p.mapExpressions { expr =>
+      rewriteWithExprAndInputPlans(expr, inputPlans)
+    }
+    newPlan = newPlan.withNewChildren(inputPlans)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala:65]: the children do not change, why calling `withNewChildren`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala [hunk 2]
```diff
@@ -229,4 +236,85 @@ class RewriteWithExpressionSuite extends PlanTest {
...
+          (a + 1).as(commonExpr2Name)): _*)
+        .groupBy($"$groupingExprName")(
+          ($"$groupingExprName" + 2).as("col1"),
+          count($"$commonExpr2Name" * $"$commonExpr2Name" - 3).as("col2")
+        )
+        .analyze
+    )
+  }
+
+  test("WITH expression in aggregate exprs") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala:290]: can we merge the two tests and have WITH in both grouping and aggregate expressions?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala [hunk 3]
```diff
@@ -229,4 +236,85 @@ class RewriteWithExpressionSuite extends PlanTest {
...
+    val Seq(a, b) = testRelation.output
+    val expr1 = With.create((a + 1, 0)) { case Seq(ref) =>
+      ref * ref
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala:293]: can we test aggregate function as the common expression?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala [hunk 4]
```diff
@@ -21,36 +21,57 @@ import scala.collection.mutable
...
-    plan.transformDownWithSubqueriesAndPruning(_.containsPattern(WITH_EXPRESSION)) {
+    plan.transformUpWithSubqueriesAndPruning(_.containsPattern(WITH_EXPRESSION)) {
+      case p @ PhysicalAggregation(
+          groupingExpressions, aggregateExpressions, resultExpressions, child)
+          if p.expressions.exists(_.containsPattern(WITH_EXPRESSION)) =>
+        // For aggregates, separate computation of the aggregations themselves from the final
+        // result by moving the final result computation into a projection above. This prevents
+        // this rule from producing an invalid Aggregate operator.
+        // TODO: the names of these aliases will become outdated after the rewrite
+        val aggExprs = aggregateExpressions.map(ae => Alias(ae, ae.toString)(ae.resultId))
```
[the remark this answers is not in the document]
Comment (kelvinjian-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala:49]: good idea, i'll change this

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala [hunk 5]
```diff
@@ -29,25 +29,29 @@ import org.apache.spark.sql.types.IntegerType
...
 
   private val testRelation = LocalRelation($"a".int, $"b".int)
   private val testRelation2 = LocalRelation($"x".int, $"y".int)
 
   test("simple common expression") {
     val a = testRelation.output.head
-    val commonExprDef = CommonExpressionDef(a)
-    val ref = new CommonExpressionRef(commonExprDef)
-    val plan = testRelation.select(With(ref + ref, Seq(commonExprDef)).as("col"))
+    val expr = With.create((a, 0)) { case Seq(ref) =>
```
[the remark this answers is not in the document]
Comment (kelvinjian-db) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala:42]: good point, i think it doesn't, i'll remove these

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala (hunk 2 above)
Comment (kelvinjian-db) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala:290]: doesn't the test above test WITH in both grouping and aggregate expressions? the test here is for testing the motivating example mentioned in https://issues.apache.org/jira/browse/SPARK-47839

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala (hunk 3 above)
Comment (kelvinjian-db) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala:293]: yes let me add a test for that

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (kelvinjian-db) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala]: i ended up changing the tests to manually extract the common expression IDs so we don't need `With.create` anymore

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala [hunk 6]
```diff
@@ -21,36 +21,68 @@ import scala.collection.mutable
...
+    plan.transformUpWithSubqueriesAndPruning(_.containsPattern(WITH_EXPRESSION)) {
+      // For aggregates, separate the computation of the aggregations themselves from the final
+      // result by moving the final result computation into a projection above it. This prevents
+      // this rule from producing an invalid Aggregate operator.
+      case p @ PhysicalAggregation(
+          groupingExpressions, aggregateExpressions, resultExpressions, child)
+          if p.expressions.exists(_.containsPattern(WITH_EXPRESSION)) =>
+        // There should not be dangling common expression references in the aggregate expressions.
+        // This can happen if a With is created with an aggregate function in its child.
+        assert(!aggregateExpressions.exists(ae =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala:50]: Shall we do the assert in the constructor of `With`, to fail earlier?

**Comment by cloud-fan** [comment]
The test fails: `org.apache.spark.sql.connect.ProtoToParsedPlanTestSuite`

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**